### ISSUE TITLE
SimpleSchema custom validation and current document ID

### ISSUE THREAD

**Issue by rantav** [aldeed/meteor-simple-schema#268]
Hi, this is part-question, part-feature request. [I asked this question on the forum](https://groups.google.com/forum/#!searchin/meteor-talk/vanity/meteor-talk/IcxGGEeaDDI/Wc9AevbGgugJ) but since I didn't get a reply let my ask it once again here...

In short - **when running custom validation on the server, is there a way to access the current document ID?** (the document being validated currently, when the operation is an update)

If not - then would you be able to advise an alternative to my approach? 
And/Or please consider this as a feature suggestion.
Here's the full question from the forum, along with the use case motivation.

I implemented a custom validation function and in this function I need to access the current document ID. Is that possible? 
If it's an insert (new document) - the document ID should be null, but if it's an update - I'd like to use the document ID in order to run the custom validation.
According to the doc, the document ID doesn't seem to be provided (https://github.com/aldeed/meteor-simple-schema#custom-validation)
Here's my scenario, maybe there's another way around this.
I have two types of collections: Teams and Companies. A team belongs to a single company (team.companyId) and a company may have many teams. The app is here btw in case you want to get the hang of it: http://teams.gormim.com

Each team has a vanityUrl for example "google/gmail" or "google/search"
Each company also has a vanity URL, for example "google" or "microsoft".

When validating the team's vanityUrl field I'd like to make sure that the URL is unique, that it isn't used yet. 
This is fine with new teams (or new companies) as the object gets inserted, I run a query over the DB and check if any other team already is using this vanity URL. 
But for updates - if I run this query, the result includes this current team's vanity URL, so in cases where the vanity URL isn't changed (e.g. other fields of the document are changed) I get an error - since vanity URL already exists, it is this document's vanity URL. 
If I had access to the current document ID (the current document that's being validated) I'd be able to exclude this result, but not having access to the current document ID - I'm at loss... 

On top of that, there are some business rules regarding the relationships b/w a team and a company, so for example it's OK to have a team's vanity URL "similar" to a company's URL in some sense (the precise logic isn't important). 
Here again - if I had the team ID when validating team.vanityUrl that would be great b/c from the teamId I can easily query the DB (server side of course) and find out the company.

So to sum up - I wish that using SimpleSchema's custom validation, I'd have access to the current document ID, at least when we're talking about updates. 
Is this possible? 

thanks!

**Comment by aldeed** [comment]
`this.docId` should be set if collection2 package is handling the update.

**Comment by rantav** [comment]
Hmm, so simple, how did I miss that? 

But I noticed that while on the **server** `this.docId` is valid, on the **client** it's `undefined`. Is this by design?

**Comment by aldeed** [comment]
For an insert? Client validation probably runs before the `mongo` pkg code that generates the _id. But it should be set for an update.

**Comment by rantav** [comment]
No, it's actually for an update, not an insert and this.docId is undefined on the client. 
So you're saying it is supposed to be set on the client as well?
OK, I'll try to get a repro sometime to send you (or find our where I'm wrong :wink:)

**Comment by aldeed** [comment]
Yes https://github.com/aldeed/meteor-collection2#custom

**Comment by rantav** [comment]
OK, repro isn't easy, on a barebones meteor project I was not yet able to repro this, but on my main repo (large and a private one) I'm able to consistently reproduce this. 
The behavior is that on the client side `this.docId` is undefined. On the server `this.docId` is fine (well defined).

However, while working to repro this on a minimal barebones project, I went into the main project and upgraded the world. (meteor to 1.0.3.1 etc) and while doing to I also upgraded simple-schema related packages, which surprisingly fixed this bug for me!

Here's the list of upgrades related to simple schema (one of them probably fixes my problem):

```
aldeed:autoform-select2@1.0.3 -> aldeed:autoform-select2@1.0.4
aldeed:collection2@2.3.0 -> aldeed:collection2@2.3.2
aldeed:simple-schema@1.1.0 -> aldeed:simple-schema@1.3.0
```

So as far as I concern the problem is solved ;)
